py/embedded-server/deephaven_server/__init__.py

```diff
@@ -3,14 +3,93 @@
 #
 
 import importlib.metadata
+import click
+import secrets
+import signal
+import sys
+import webbrowser
 
-from .start_jvm import DEFAULT_JVM_PROPERTIES, DEFAULT_JVM_ARGS, start_jvm
 from .server import Server
 
 from deephaven_internal.jvm import check_py_env
+
 check_py_env()
 
 # Note: this is the _distribution_ name, not the _package_ name. Until 3.10, there is not an easy way to get the
 # distribution name from the package name.
 # https://docs.python.org/3/library/importlib.metadata.html#package-distributions
-__version__ = importlib.metadata.version('deephaven-server')
+__version__ = importlib.metadata.version("deephaven-server")
+
+
+@click.group()
+def cli():
+    """
+    Command-line interface entry point for the Deephaven embedded server application.
+    Accepts a command to start the application and the arguments to use.
+    """
+    pass
+
+
+@cli.command()
+@click.option("--host", default=None, help="The host to bind to.")
+@click.option("--port", default=None, type=int, help="The port to bind to.")
+@click.option(
+    "--key",
+    default=None,
+    help="Start the server in pre-shared key mode. If the key is empty, a random key will be used. Cannot be used if anonymous is specified.",
+)
+@click.option(
+    "--anonymous",
+    is_flag=True,
+    help="Start the server in anonymous mode. Cannot be used if key is specified.",
+)
+@click.option("--jvm-args", default=None, help="The JVM arguments to use.")
+def server(host, port, key, anonymous, jvm_args):
+    """
+    Start the Deephaven server.
+    """
+    click.echo("Starting Deephaven server...")
+
+    if anonymous and key is not None:
+        raise click.ClickException("Cannot specify both --anonymous and --key")
+
+    if jvm_args is None:
+        jvm_args = ""
+    jvm_args = jvm_args.split()
+
+    if anonymous:
+        jvm_args += [f"-DAuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler"]
+    elif key is not None:
+        jvm_args += [
```

Comment:
Looks like this prevents a user from specifying mTLS/etc as their preferred option, without forcing that they also have PSK (or anon) enabled as well... and it will override their `AuthHandlers` value and wipe out their own config.

Unsure if we want to get into the business of parsing AuthHandlers or other config vars here... if we want to go after this it might be time to tighten up our config language/control, or just be prepared for breaking changes.

Or put dh_args back or something, and require that it be used instead of system props for dh configs? and have some kind of schema here for "this is actually a `,`-delimited list of strings"?

I kind of hate all of the choices... I do have a draft proposal for "we should use TOML", but I'm not sure now's the time to pick that up in earnest?

Reply:
Additionally, these shouldn't be mutually exclusive - one day we'll have authorization hooks that are easy to use, and we'll want to allow both (anon might be simple, psk might be admin).

Reply:
I'm _hoping_ we can discourage users from setting `--anonymous` or `--psk` wrt a deephaven server start command (in my helper PR I originally removed them). When they are not provided, the server will retain it's normal defaults or source from the users configuration file. I also suggested that `deephaven set-config ...` might be a reasonable command that allows us to explicitly present a way to do "proper" configuration from the CLI (ie, writing down into a configuration file w/ proper safeguards).

There was a similar concern I had for jsPlugins configuration, essentially relying on a prefix call:

```java
class PropertyFile {
...
    /**
     * Collect all of the properties in this property file that begin with a given prefix.
     *
     * @return a new Properties object containing the selected properties, with the prefix removed.
     */
    public Properties getProperties(String prefix) {
```

to allow independent configuration. ie

```properties
deephaven.jsPlugins.foo=/path/to/plugin-1
deephaven.jsPlugins.bar=/path/to/plugin-2
```

as opposed to needing to do

```properties
deephaven.jsPlugins=/path/to/plugin-1,/path/to/plugin-2
```

Reply:
I don't think we want or need `deephaven server` command line options to present the full universe of configuration possibilities to users; in my mind, these options may be better thought of as `--only-psk=<key>`, `--only-anonymous`, etc. I am leaning that maybe we should remove them from `deephaven server` altogether, I don't think it would be a big loss. (I could make similar argument for `--host` and `--port`, but I think that's a slightly harder sell.)

Reply:
I'd be alright with getting rid of the anonymous/psk options now that there's the `targetUrls` option, as those options conflicted with existing config. Leaving `jvm_args` seems alright, it adds to the args in your existing config. `host`/`port` seem like they'd be nice and would just override any config you have which makes sense IMO.